File: game/scripts/vscripts/abilities/ward_slot/ability_ward_observer_slot.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__DecorateLegacy = ____lualib.__TS__DecorateLegacy
local ____exports = {}
local ____dota_ts_adapter = require("utils.dota_ts_adapter")
local BaseAbility = ____dota_ts_adapter.BaseAbility
local registerAbility = ____dota_ts_adapter.registerAbility
--- 真眼额外槽位：充能由商店购买真眼经 WardSlot filter 注入，放置时消耗一层充能。
____exports.AbilityWardObserverSlot = __TS__Class()
local AbilityWardObserverSlot = ____exports.AbilityWardObserverSlot
AbilityWardObserverSlot.name = "AbilityWardObserverSlot"
__TS__ClassExtends(AbilityWardObserverSlot, BaseAbility)
function AbilityWardObserverSlot.prototype.OnSpellStart(self)
    local caster = self:GetCaster()
    local point = self:GetCursorPosition()
    local lifetime = self:GetSpecialValueFor("lifetime")
    local ward = CreateUnitByName(
        "npc_dota_observer_wards",
        point,
        false,
        caster,
        caster,
        caster:GetTeamNumber()
    )
    ward:SetControllableByPlayer(
        caster:GetPlayerOwnerID(),
        true
    )
    ward:AddNewModifier(caster, self, "modifier_kill", {duration = lifetime})
end
AbilityWardObserverSlot = __TS__DecorateLegacy(
    {registerAbility(nil, "abilities/ward_slot/ability_ward_observer_slot")},
    AbilityWardObserverSlot
)
____exports.AbilityWardObserverSlot = AbilityWardObserverSlot
return ____exports

File: src/vscripts/abilities/ts_abilities/ward_slot/ability_ward_observer_slot.ts
import { BaseAbility, registerAbility } from '../../../utils/dota_ts_adapter';
import { isInWardNoCastZone } from './ward-no-cast-zone';

/** 参考 1x6 abilities/ui/observer.lua */
@registerAbility('ability_ward_observer_slot')
export class AbilityWardObserverSlot extends BaseAbility {
  IsRefreshable(): boolean {
    return false;
  }

  // CastFilterResultLocation 返回 UF_FAIL_CUSTOM 时引擎才会取 GetCustomCastErrorLocation 飘字，两者须配套
  CastFilterResultLocation(location: Vector): UnitFilterResult {
    return isInWardNoCastZone(location) ? UnitFilterResult.FAIL_CUSTOM : UnitFilterResult.SUCCESS;
  }

  GetCustomCastErrorLocation(): string {
    return '#dota_hud_error_ward_no_cast_zone';
  }

  // 充能只能来自购买转换；引擎在 OnStolen 后才写入偷取充能，延迟一帧清零才生效
  OnStolen(): void {
    Timers.CreateTimer(0, () => this.SetCurrentAbilityCharges(0));
  }

  OnSpellStart(): void {
    const caster = this.GetCaster();
    const point = this.GetCursorPosition();
    const lifetime = this.GetSpecialValueFor('lifetime');

    const ward = CreateUnitByName(
      'npc_dota_observer_wards',
      point,
      true,
      caster,
      caster,
      caster.GetTeamNumber(),
    );
    ward.SetControllableByPlayer(caster.GetPlayerOwnerID(), true);
    ward.SetOwner(caster);
    ward.AddNewModifier(caster, undefined, 'modifier_item_buff_ward', {
      duration: lifetime,
    });
    ward.AddNewModifier(caster, this, 'modifier_kill', { duration: lifetime });
  }
}
